tests/test_oml/test_samplers/test_balance.py

```diff
@@ -66,64 +66,46 @@ def input_for_balance_batch_sampler() -> TLabelsPK:
-        n_labels: Number of labels in a batch
-        n_instances: Number of instances for each label in a batch
-
-    """
+def check_balance_batch_sampler_epoch(sampler: BalanceSampler, labels: List[int]) -> None:
     sampled_ids = list(sampler)
-
-    sampled_labels = []
     collected_labels: Set[int] = set()
+
     # emulating of 1 epoch
     for i, batch_ids in enumerate(sampled_ids):
         batch_labels = itemgetter(*batch_ids)(labels)  # type: ignore
-        assert all(label not in collected_labels for label in batch_labels)
-        collected_labels.update(batch_labels)
 
         labels_counter = Counter(batch_labels)
         num_batch_labels = len(labels_counter)
-        num_batch_samples = list(labels_counter.values())
-        cur_batch_size = len(batch_labels)
-        sampled_labels.extend(list(labels_counter.keys()))
+        num_batch_samples_counts = list(labels_counter.values())
 
         # batch-level invariants
+        assert all(label not in collected_labels for label in batch_labels)
         assert len(set(batch_ids)) >= 4, set(batch_ids)  # type: ignore
+        assert len(batch_ids) == sampler.n_labels * sampler.n_instances
+        assert num_batch_labels == sampler.n_labels, (num_batch_labels, sampler.n_labels)
+        assert all(el == sampler.n_instances for el in num_batch_samples_counts)
 
-        is_last_batch = i == sampler.batches_in_epoch - 1
-        if is_last_batch:
-            assert 1 < num_batch_labels <= n_labels
-            assert all(1 < el <= n_instances for el in num_batch_samples)
-            assert 2 * 2 <= cur_batch_size <= n_labels * n_instances
-        else:
-            assert num_batch_labels == n_labels, (num_batch_labels, n_labels)
-            assert all(el == n_instances for el in num_batch_samples)
-            assert cur_batch_size == n_labels * n_instances
+        collected_labels.update(batch_labels)
 
     # epoch-level invariants
+    n_expected_batches = len(set(labels)) // sampler.n_labels
     num_labels_in_data = len(set(labels))
-    num_labels_in_sampler = len(set(sampled_labels))
-    assert (num_labels_in_data == num_labels_in_sampler) or (num_labels_in_data == num_labels_in_sampler + 1)
-
+    num_labels_sampled = len(collected_labels)
     n_instances_sampled = sum(map(len, sampled_ids))  # type: ignore
-    assert (num_labels_in_data - 1) * n_instances <= n_instances_sampled <= num_labels_in_data * n_instances, (
-        n_instances_sampled,
-        num_labels_in_data * n_instances,
-    )
+
+    assert len(sampler) == n_expected_batches, (len(sampler), n_expected_batches)
+    assert (num_labels_in_data - sampler.n_labels + 1) <= num_labels_sampled <= num_labels_in_data
+    assert len(sampler) * sampler.n_labels * sampler.n_instances == n_instances_sampled
 
 
-def test_balance_batch_sampler(input_for_balance_batch_sampler: TLabelsPK) -> None:
+def test_balance_batch_sampler(input_for_balance_batch_sampler: TLabels) -> None:
```

Comment:
[suggestion] I prefer to launch tests with some parametrize, e.g. generate cases with different num of labels. Sometimes it generate you some cases with dividing, rounding, etc. 

Up to you